taichi/program/async_engine.h

```diff
@@ -19,6 +19,62 @@ TLANG_NAMESPACE_BEGIN
 
 // TODO(yuanming-hu): split into multiple files
 
+class IRHandle {
+ public:
+  IRHandle(IRNode const *ir, uint64 hash) : ir_(ir), hash_(hash) {
+  }
+
+  std::unique_ptr<IRNode> clone() const;
+
+  IRNode const *ir() const {
+    return ir_;
+  }
+
+  uint64 hash() const {
+    return hash_;
+  }
+
+  // Two IRHandles are considered the same iff their hash values are the same.
+  bool operator==(const IRHandle &other_ir_handle) const {
+    return hash_ == other_ir_handle.hash_;
+  }
+
+ private:
+  IRNode const *ir_;  // not owned
+  uint64 hash_;
+};
+
+TLANG_NAMESPACE_END
+
+namespace std {
+template <>
+struct hash<taichi::lang::IRHandle> {
+  std::size_t operator()(taichi::lang::IRHandle const &ir_handle) const
+      noexcept {
+    return ir_handle.hash();
+  }
+};
+}  // namespace std
+
+TLANG_NAMESPACE_BEGIN
+
+class IRBank {
+ public:
+  std::vector<std::unique_ptr<IRNode>> trash_bin;
```

Comment:
Nit: I'd suggest to make `trashi_bin` private, and add a `insert_to_trash_bin()` method for these reasons:

1. `insert()` could use `insert_to_trash_bin()`
2. In `ExecutionQueue::enqueue`, we can also use this method to collect the cloned `stmt` (provided that ExecutionQueue has access to IRBank..)

WDYT?